File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        int i=0,j=0;
        unordered_map<char,int>mp;
        int len=0;
        for(j=0;j<n;++j){
            if(mp.find(s[j])!=mp.end()){
                i=max(i,mp[s[j]]+1);
            }
            mp[s[j]]=j;

            len=max(len,j-i+1);
        }
        return len;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer>hMap=new HashMap<>();
        int n=s.length();
        int maxLen=0;
        int j=0;
        for(int i=0;i<n;++i){
            if(hMap.containsKey(s.charAt(i))){
                j=Math.max(hMap.get(s.charAt(i))+1,j);
                // maxLen=Math.max(maxLen,i-hMap.get(s.charAt(i)));
            }
            maxLen=Math.max(maxLen,i-j+1);
            hMap.put(s.charAt(i),i);
        }
        return maxLen;
    }
}